### PR TITLE
chore: Use reference to an array representing a blob instead of an owned KzgBlob (only for peerdas functions)

### PR DESCRIPTION
**Pull request by kevaundray**
## Issue Addressed

This changes the blob type in the kzg inteface to take a reference to a Blob instead of taking ownership of a Blob. We are still allocating the blob on the stack inside of the library.

## Proposed Changes

Please list or describe the changes introduced by this PR.

## Additional Info

Please provide any additional information. For example, future considerations
or information useful for reviewers.

### CODE REVIEW

File: beacon_node/beacon_chain/src/kzg_utils.rs [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 use types::{Blob, DataColumnSidecar, EthSpec, Hash256, KzgCommitment, KzgProof};
 
 /// Converts a blob ssz List object to an array to be used with the kzg
 /// crypto library.
-fn ssz_blob_to_crypto_blob<E: EthSpec>(blob: &Blob<E>) -> Result<KzgBlob, KzgError> {
-    KzgBlob::from_bytes(blob.as_ref()).map_err(Into::into)
+fn ssz_blob_to_crypto_blob<E: EthSpec>(blob: &Blob<E>) -> Result<KzgBlobRef, KzgError> {
+    let blob_bytes: &[u8] = blob.as_ref();
+    Ok(blob_bytes.try_into().expect("expected cell to have size {BYTES_PER_BLOB}. This should be guaranteed by the `FixedVector type"))
 }
```
Comment (kevaundray) [inline beacon_node/beacon_chain/src/kzg_utils.rs]: The main change is here, we push the allocation of the blob into the kzg library

File: crypto/kzg/src/lib.rs [hunk 2]
```diff
@@ -173,7 +189,11 @@ impl Kzg {
     }
 
     /// Computes the cells and associated proofs for a given `blob` at index `index`.
-    pub fn compute_cells_and_proofs(&self, blob: &Blob) -> Result<CellsAndKzgProofs, Error> {
+    #[allow(clippy::needless_lifetimes)]
+    pub fn compute_cells_and_proofs<'a>(
+        &self,
+        blob: KzgBlobRef<'a>,
+    ) -> Result<CellsAndKzgProofs, Error> {
         let blob_bytes: &[u8; BYTES_PER_BLOB] = blob
```
Comment (jimmygchen) [inline crypto/kzg/src/lib.rs]: I think this conversion can now be removed as `KzgBlobRef` is the exact same type?

File: consensus/types/src/data_column_sidecar.rs [hunk 3]
```diff
@@ -128,8 +128,11 @@ impl<E: EthSpec> DataColumnSidecar<E> {
         let blob_cells_and_proofs_vec = blobs
             .into_par_iter()
             .map(|blob| {
-                let blob = KzgBlob::from_bytes(blob).map_err(KzgError::from)?;
-                kzg.compute_cells_and_proofs(&blob)
+                let blob = blob
+                    .as_ref()
+                    .try_into()
+                    .expect("blob should have a guaranteed size due to FixedVector");
+                kzg.compute_cells_and_proofs(blob)
```
Comment (jimmygchen) [inline consensus/types/src/data_column_sidecar.rs:135]: I think this is the main change that fixes the stack overflow on `das` testing right?

The other changes change mainnet code path - it's probably not something we can merge to the `unstable` branch now, although I think we do want them eventually when `c-kzg` adopts it or when `rust-eth-kzg` is production-ready for mainnet.

I think we can merge this branch to `das` and I'll just cherry-pick the changes relevant to `compute_cells_and_proofs` onto `unstable`.

File: beacon_node/beacon_chain/src/kzg_utils.rs (hunk 1 above)
Comment (jimmygchen) [reply beacon_node/beacon_chain/src/kzg_utils.rs]: Cool, could we adopt the same change for `c-kzg-4844`? (to prepare it for increased blob count)

**Review by jimmygchen — CHANGES_REQUESTED** [review]
Looks good thanks @kevaundray , please see my comments above!
We can probably merge this to `das` if you agree, and I'll create a PR to merge to `unstable` soon.

File: consensus/types/src/data_column_sidecar.rs (hunk 3 above)
Comment (kevaundray) [reply consensus/types/src/data_column_sidecar.rs:135]: Yep this is correct

File: crypto/kzg/src/lib.rs (hunk 2 above)
Comment (kevaundray) [reply crypto/kzg/src/lib.rs]: Oh yep that's true, this conversion is infallible

File: beacon_node/beacon_chain/src/kzg_utils.rs (hunk 1 above)
Comment (kevaundray) [reply beacon_node/beacon_chain/src/kzg_utils.rs]: I think c-kzg would need to allocate under the hood -- I'll open up a PR when I get back on my laptop to track arguments for and against

**Comment by kevaundray** [comment]
> Looks good thanks @kevaundray , please see my comments above!
> 
> We can probably merge this to `das` if you agree, and I'll create a PR to merge to `unstable` soon.

Yep agreed to all comments -- I'm away from keyboard at the moment, so feel free to cherry-pick changes without me

**Review by jimmygchen — APPROVED** [review]
I've update the PR. Thanks again for this @kevaundray !

**Pull request merged**